build-tests/api-extractor-scenarios/etc/test-outputs/ambientNameConflict/api-extractor-scenarios.api.json

```diff
@@ -62,7 +63,8 @@
             },
             {
               "kind": "Reference",
-              "text": "MyPromise"
+              "text": "MyPromise",
+              "canonicalReference": "api-extractor-scenarios!Promise:class"
```

Comment:
*Issue: No reference should be generated because the `api-extractor-scenarios` package does not export a symbol `Promise`.*

Reply:
The problem is that it imports `Promise` from `./localFile` as the alias `MyPromise`. I get the name `Promise` by resolving the alias.

Perhaps this should have been `"api-extractor-scenarios/localFile!Promise:class"`? The problem with that is that api-extractor doesn't currently seem to support anything other than package-level exports.

Also, you include the definition of `MyPromise` as an unexported class in the rollup.d.ts named `Promise_2`, so perhaps this should have been `"api-extractor-scenarios!~Promise_2:class"`?

Reply:
Also, the `ExcerptBuilder` can't know whether the custom `Promise` here might be publicly exported as some completely different name by the package.

Reply:
@rbuckton In @adventure-yunfei's PR https://github.com/microsoft/web-build-tools/pull/1005 (an earlier attempt at this), the approach was to build a [`Map<ts.Symbol, ApiItem>`](https://github.com/microsoft/web-build-tools/blob/f53c62accc4b0154a3b2a3ab00aa3a6ca7a8e812/apps/api-extractor/src/generators/ApiModelGenerator.ts#L48) for all exported APIs, and then consult that when generating the excerpt references.   (He didn't consider how to choose among multiple exports, but a heuristic would be reasonable.)

> Also, you include the definition of `MyPromise` as an unexported class in the rollup.d.ts named `Promise_2`, so perhaps this should have been `"api-extractor-scenarios!~Promise_2:class"`?

As far as I can see, the purpose of these references is for linking to documentation entries.  As long as API Extractor does not generate documentation for unexported symbols, then there isn't much benefit in referring to them.  The `api-extractor-scenarios.api.json` captures the complete inventory of possible documentation entries, we based on that we should be able to determine that `api-extractor-scenarios!Promise:class` isn't linkable, and thus should be omitted.

I'd like to get your PR merged though, so maybe we should postpone this issue and tackle it in a subsequent PR (unless it's a very simple addition).

Reply:
> The api-extractor-scenarios.api.json captures the complete inventory of possible documentation entries

I've mentioned before that this is an incorrect assumption. DocFX supports something called an `xrefmap.yml` file that lets you link to documentation *not* included in your project. This is useful if you are building an API on top of an external API and you want to link to the external API's documentation as well.

Reply:
To clarify, I was referring to the subset of links that refer to declarations defined by the current package. For example, the `ts.SourceFile` containing the (followed) declaration for `MyPromise` is part of the file set for the current package, thus we can be sure there is no documentation entry for it.  There's no way it could appear in `xrefmap.yml`, right? 

For references to symbols imported from other packages, we should preserve those in case they are documented as you say. BTW if an external package publishes a [tsdoc-metadata.json](https://github.com/microsoft/tsdoc/issues/7#issuecomment-442271815) file, then we could possibly prove that its declarations would not appear in `xrefmap.yml` also, I think.

Does that make sense?

Reply:
> To clarify, I was referring to the subset of links that refer to declarations defined by the current package. 

What is the action item here then? Do I need to make a change in api-extractor to skip adding a `canonicalReference` to an excerpt for a non-exported member in the same package? Its unfortunate that those types will be printed in the excerpt but unreachable in the documentation. Should I be introducing a symbol -> ApiItem mapping as you mentioned to verify this, or some other mechanism?

> BTW if an external package publishes a tsdoc-metadata.json file, then we could possibly prove that its declarations would not appear in `xrefmap.yml` also, I think.

I suppose it does, though I'm not sure I want to take on the added complexity of that in this PR.

Reply:
@rbuckton I think so, but let's postpone that for a later PR. I don't see any practical problems with the current implementation. As mentioned on Gitter, tomorrow we'll ask SharePoint to validate against their full data set, and if there are no issues then we should be good to finally merge it. (Hurray!)

BTW I'm really excited about this feature! Thanks so much for finishing it. After the PR is merged, I want to get it enabled for markdown output as well.